File: fetch_toyota_xbrl.py
#!/usr/bin/env python3
"""
EDINET API v2 - トヨタ自動車の有価証券報告書から当期純利益・発行済株式数を取得するスクリプト
"""

import os
import time
import zipfile
import io
from datetime import date, timedelta

import requests
from lxml import etree

EDINET_API_KEY = os.environ.get("EDINET_API_KEY", "")
BASE_URL = "https://api.edinet-fsa.go.jp/api/v2"

TOYOTA_EDINET_CODE = "E02144"

# 3月期決算なので有報提出は6月頃。直近数年の6月を候補にする
CANDIDATE_DATES = [
    date(2024, 6, 25),
    date(2024, 6, 24),
    date(2024, 6, 26),
    date(2024, 6, 27),
    date(2024, 6, 28),
    date(2023, 6, 26),
    date(2023, 6, 27),
    date(2023, 6, 28),
]


def check_api_key():
    if not EDINET_API_KEY:
        print("[ERROR] 環境変数 EDINET_API_KEY が設定されていません。")
        print("  export EDINET_API_KEY='your_api_key' を実行してください。")
        raise SystemExit(1)


def get_params(**kwargs):
    params = {"Subscription-Key": EDINET_API_KEY}
    params.update(kwargs)
    return params


def find_toyota_doc_id():
    """書類一覧APIを使ってトヨタ自動車の有価証券報告書docIDを探す"""
    print("=== トヨタ自動車の有報docIDを検索中 ===")

    for target_date in CANDIDATE_DATES:
        date_str = target_date.strftime("%Y-%m-%d")
        print(f"\n[検索] 提出日: {date_str}")

        params = get_params(date=date_str, type=2)
        resp = requests.get(f"{BASE_URL}/documents.json", params=params, timeout=30)

        if resp.status_code != 200:
            print(f"  [ERROR] ステータス: {resp.status_code}")
            print(f"  レスポンス: {resp.text[:300]}")
            time.sleep(3)
            continue

        data = resp.json()
        results = data.get("results", []) or []

        for doc in results:
            edinet_code = doc.get("edinetCode", "")
            ordinance = doc.get("ordinanceCode", "")
            form = doc.get("formCode", "")
            filer = doc.get("filerName", "")
            doc_id = doc.get("docID", "")

            if (
                edinet_code == TOYOTA_EDINET_CODE
                and ordinance == "010"
                and form == "030000"
            ):
                print(f"  [発見] filerName={filer}, docID={doc_id}, 提出日={date_str}")
                return doc_id, date_str

        print(f"  該当なし（{len(results)}件中）")
        time.sleep(3)

    print("\n[ERROR] 対象期間内にトヨタ自動車の有報が見つかりませんでした。")
    print("CANDIDATE_DATES を調整して再試行してください。")
    raise SystemExit(1)


def download_xbrl_zip(doc_id):
    """書類取得API（type=1）でXBRL ZIPをダウンロード"""
    print(f"\n=== XBRLダウンロード (docID={doc_id}) ===")

    params = get_params(type=1)
    url = f"{BASE_URL}/documents/{doc_id}"
    resp = requests.get(url, params=params, timeout=120, stream=True)

    if resp.status_code != 200:
        print(f"[ERROR] ステータス: {resp.status_code}")
        print(f"レスポンス: {resp.text[:300]}")
        raise SystemExit(1)

    content = resp.content
    print(f"  ダウンロード完了: {len(content) / 1024:.1f} KB")
    return content


def extract_xbrl_from_zip(zip_bytes):
    """ZIPからXBRLファイルを取り出す"""
    print("\n=== ZIPを解凍してXBRLファイルを探す ===")

    with zipfile.ZipFile(io.BytesIO(zip_bytes)) as zf:
        names = zf.namelist()
        xbrl_files = [n for n in names if n.lower().endswith(".xbrl")]

        print(f"  ZIP内ファイル数: {len(names)}")
        print(f"  XBRLファイル: {xbrl_files}")

        if not xbrl_files:
            print("[ERROR] ZIPにXBRLファイルが見つかりません。")
            print("ZIP内容:", names[:20])
            raise SystemExit(1)

        # 複数ある場合はパスが短い（ルートに近い）ものを優先
        target = sorted(xbrl_files, key=lambda x: len(x))[0]
        print(f"  使用するXBRL: {target}")
        return zf.read(target)


def parse_xbrl(xbrl_bytes):
    """XBRLをパースして当期純利益・発行済株式数を抽出"""
    print("\n=== XBRLをパース ===")

    root = etree.fromstring(xbrl_bytes)

    # 名前空間マップを収集
    ns_map = {}
    for elem in root.iter():
        ns_map.update(elem.nsmap)

    # 全タグ名（ローカル名）を収集
    all_local_names = set()
    for elem in root.iter():
        local = etree.QName(elem.tag).localname if "}" in elem.tag else elem.tag
        all_local_names.add(local)

    # 候補タグのキーワード
    profit_keywords = [
        "NetIncomeLoss",
        "ProfitLossAttributableToOwnersOfParent",
        "ProfitLoss",
        "NetIncome",
        "profit",
        "loss",
        "Income",
    ]
    shares_keywords = [
        "NumberOfIssuedShares",
        "IssuedShares",
        "TotalNumberOfIssuedShares",
        "shares",
        "Shares",
        "株式数",
    ]

    print("\n--- 純利益関連タグ候補 ---")
    profit_candidates = sorted(
        [t for t in all_local_names if any(k.lower() in t.lower() for k in profit_keywords)]
    )
    for t in profit_candidates:
        print(f"  {t}")

    print("\n--- 発行済株式数関連タグ候補 ---")
    shares_candidates = sorted(
        [t for t in all_local_names if any(k.lower() in t.lower() for k in shares_keywords)]
    )
    for t in shares_candidates:
        print(f"  {t}")

    # 値を抽出するヘルパー
    def find_value(keywords, context_hint="Duration"):
        """優先度順にタグを探し、contextRefがDurationまたはInstantのものを返す"""
        for keyword in keywords:
            for elem in root.iter():
                local = etree.QName(elem.tag).localname if "}" in elem.tag else elem.tag
                if keyword.lower() in local.lower():
                    context_ref = elem.get("contextRef", "")
                    text = (elem.text or "").strip()
                    if text and text.lstrip("-").isdigit():
                        return local, text, context_ref
        return None, None, None

    print("\n=== 抽出結果 ===")

    # 当期純利益
    profit_priority = [
        "ProfitLossAttributableToOwnersOfParent",
        "NetIncomeLoss",
        "ProfitLoss",
        "NetIncome",
    ]
    tag, value, ctx = find_value(profit_priority)
    if value:
        val_million = int(value) // 1_000_000 if abs(int(value)) >= 1_000_000 else int(value)
        print(f"\n当期純利益:")
        print(f"  タグ名     : {tag}")
        print(f"  contextRef : {ctx}")
        print(f"  値（原単位）: {int(value):,}")
        # EDINETのXBRLは通常「円」単位または「百万円」単位
        # decimalsで確認
        for elem in root.iter():
            local = etree.QName(elem.tag).localname if "}" in elem.tag else elem.tag
            if tag and tag.lower() in local.lower() and (elem.text or "").strip() == value:
                decimals = elem.get("decimals", "不明")
                unit_ref = elem.get("unitRef", "不明")
                print(f"  decimals   : {decimals}")
                print(f"  unitRef    : {unit_ref}")
                # decimals="-6" → 百万円単位, decimals="0" → 円単位
                if decimals == "-6":
                    print(f"\n当期純利益：{int(value):,} 百万円")
                elif decimals == "0" or decimals == "-3":
                    divisor = 1 if decimals == "0" else 1_000
                    print(f"\n当期純利益：{int(value) // (1_000_000 // divisor):,} 百万円（{int(value):,} 円換算）")
                else:
                    print(f"\n当期純利益：{int(value):,}（単位はdecimalsを確認）")
                break
    else:
        print("\n当期純利益: タグが見つかりませんでした。")
        print("上記の「純利益関連タグ候補」からタグ名を確認してください。")

    # 発行済株式数
    shares_priority = [
        "NumberOfIssuedSharesSummaryOfBusinessResults",
        "TotalNumberOfIssuedShares",
        "NumberOfIssuedShares",
        "IssuedShares",
    ]
    tag_s, value_s, ctx_s = find_value(shares_priority, context_hint="Instant")
    if value_s:
        print(f"\n発行済株式数:")
        print(f"  タグ名     : {tag_s}")
        print(f"  contextRef : {ctx_s}")
        print(f"  値         : {int(value_s):,} 株")
        print(f"\n発行済株式数：{int(value_s):,} 株")
    else:
        print("\n発行済株式数: タグが見つかりませんでした。")
        print("上記の「発行済株式数関連タグ候補」からタグ名を確認してください。")


def main():
    check_api_key()

    doc_id, submit_date = find_toyota_doc_id()
    time.sleep(4)

    zip_bytes = download_xbrl_zip(doc_id)
    time.sleep(3)

    xbrl_bytes = extract_xbrl_from_zip(zip_bytes)

    parse_xbrl(xbrl_bytes)


if __name__ == "__main__":
    main()

File: fetch_toyota_xbrl.R
# EDINET API v2 - トヨタ自動車の有価証券報告書から当期純利益・発行済株式数を取得
# 必要パッケージ: httr2, xml2
# install.packages(c("httr2", "xml2"))

library(httr2)
library(xml2)

# ---------------------------------------------------------------------------
# 設定
# ---------------------------------------------------------------------------
EDINET_API_KEY  <- Sys.getenv("EDINET_API_KEY")
BASE_URL        <- "https://api.edinet-fsa.go.jp/api/v2"
TOYOTA_EDINET   <- "E02144"

# 3月期決算 → 有報提出は6月頃。複数日付を候補にする
CANDIDATE_DATES <- c(
  "2024-06-25", "2024-06-24", "2024-06-26", "2024-06-27", "2024-06-28",
  "2023-06-26", "2023-06-27", "2023-06-28"
)

# ---------------------------------------------------------------------------
# ヘルパー
# ---------------------------------------------------------------------------
check_api_key <- function() {
  if (nchar(EDINET_API_KEY) == 0) {
    stop(paste(
      "環境変数 EDINET_API_KEY が設定されていません。\n",
      "  Sys.setenv(EDINET_API_KEY = 'your_api_key')  または\n",
      "  .Renviron に EDINET_API_KEY=your_key を追記してください。"
    ))
  }
}

edinet_get <- function(path, ...) {
  req <- request(paste0(BASE_URL, path)) |>
    req_url_query(`Subscription-Key` = EDINET_API_KEY, ...) |>
    req_timeout(120) |>
    req_error(is_error = \(resp) FALSE)   # エラーも手動でハンドル

  resp <- req_perform(req)

  if (resp_status(resp) != 200) {
    cat(sprintf("[ERROR] %s → HTTP %d\n", path, resp_status(resp)))
    cat(substr(resp_body_string(resp), 1, 300), "\n")
    stop("API request failed")
  }
  resp
}

# ---------------------------------------------------------------------------
# Step 1: docID を特定
# ---------------------------------------------------------------------------
find_toyota_doc_id <- function() {
  cat("=== トヨタ自動車の有報 docID を検索中 ===\n")

  for (d in CANDIDATE_DATES) {
    cat(sprintf("\n[検索] 提出日: %s\n", d))

    resp <- tryCatch(
      edinet_get("/documents.json", date = d, type = 2),
      error = function(e) { cat("  リクエスト失敗:", conditionMessage(e), "\n"); NULL }
    )
    if (is.null(resp)) { Sys.sleep(3); next }

    body    <- resp_body_json(resp)
    results <- body$results

    if (length(results) == 0) {
      cat(sprintf("  該当なし（0件）\n"))
      Sys.sleep(3)
      next
    }

    for (doc in results) {
      if (
        identical(doc$edinetCode,    TOYOTA_EDINET) &&
        identical(doc$ordinanceCode, "010")         &&
        identical(doc$formCode,      "030000")
      ) {
        cat(sprintf("  [発見] filerName=%s  docID=%s\n", doc$filerName, doc$docID))
        return(list(doc_id = doc$docID, submit_date = d))
      }
    }

    cat(sprintf("  該当なし（%d件中）\n", length(results)))
    Sys.sleep(3)
  }

  stop("対象期間内にトヨタ自動車の有報が見つかりませんでした。CANDIDATE_DATES を調整してください。")
}

# ---------------------------------------------------------------------------
# Step 2: XBRL ZIP をダウンロード
# ---------------------------------------------------------------------------
download_xbrl_zip <- function(doc_id) {
  cat(sprintf("\n=== XBRL ダウンロード (docID=%s) ===\n", doc_id))

  resp <- edinet_get(sprintf("/documents/%s", doc_id), type = 1)

  raw_bytes <- resp_body_raw(resp)
  cat(sprintf("  ダウンロード完了: %.1f KB\n", length(raw_bytes) / 1024))
  raw_bytes
}

# ---------------------------------------------------------------------------
# Step 3: ZIP を解凍して XBRL バイト列を返す
# ---------------------------------------------------------------------------
extract_xbrl_from_zip <- function(raw_bytes) {
  cat("\n=== ZIP を解凍して XBRL ファイルを探す ===\n")

  tmp_zip  <- tempfile(fileext = ".zip")
  tmp_dir  <- tempdir()
  on.exit(unlink(tmp_zip))

  writeBin(raw_bytes, tmp_zip)
  all_files <- unzip(tmp_zip, list = TRUE)$Name
  cat(sprintf("  ZIP 内ファイル数: %d\n", length(all_files)))

  xbrl_files <- all_files[grepl("\\.xbrl$", all_files, ignore.case = TRUE)]
  cat(sprintf("  XBRL ファイル一覧:\n"))
  cat(paste("   ", xbrl_files, collapse = "\n"), "\n")

  if (length(xbrl_files) == 0) {
    cat("ZIP 内容:\n"); print(head(all_files, 20))
    stop("XBRL ファイルが見つかりませんでした。")
  }

  # PublicDoc 配下の有価証券報告書本体を優先。AuditDoc（監査報告書）は除外する。
  public_xbrl <- xbrl_files[
    grepl("PublicDoc", xbrl_files) &
    !grepl("AuditDoc", xbrl_files) &
    grepl("jpcrp030000-asr-001", basename(xbrl_files))
  ]

  if (length(public_xbrl) == 0) {
    # jpcrp030000-asr-001 に限定せず PublicDoc 配下全体にフォールバック
    public_xbrl <- xbrl_files[
      grepl("PublicDoc", xbrl_files) & !grepl("AuditDoc", xbrl_files)
    ]
  }

  target <- if (length(public_xbrl) > 0) public_xbrl[1] else xbrl_files[1]
  cat(sprintf("  使用する XBRL: %s\n", target))

  unzip(tmp_zip, files = target, exdir = tmp_dir, overwrite = TRUE)
  readBin(file.path(tmp_dir, target), "raw", n = 100e6)
}

# ---------------------------------------------------------------------------
# Step 4: XBRL をパースして値を抽出
# ---------------------------------------------------------------------------
parse_xbrl <- function(xbrl_bytes) {
  cat("\n=== XBRL をパース ===\n")

  doc  <- read_xml(xbrl_bytes)
  ns   <- xml_ns(doc)                      # 名前空間一覧
  all_nodes <- xml_find_all(doc, "//*")    # 全要素

  # ローカル名（名前空間プレフィックスなし）を取得
  local_names <- xml_name(all_nodes, ns)   # "prefix:localname" 形式
  # ":" 以降だけ取り出す
  local_only  <- sub("^[^:]+:", "", local_names)

  # --- タグ候補を表示 -------------------------------------------------------
  profit_kw <- c("NetIncomeLoss", "ProfitLoss", "NetIncome", "profit", "loss", "Income")
  shares_kw <- c("NumberOfIssuedShares", "IssuedShares", "TotalNumberOfIssued", "Shares", "shares")

  profit_tags <- unique(local_only[Reduce(`|`, lapply(profit_kw, grepl, x = local_only, ignore.case = TRUE))])
  shares_tags <- unique(local_only[Reduce(`|`, lapply(shares_kw, grepl, x = local_only, ignore.case = TRUE))])

  cat("\n--- 純利益関連タグ候補 ---\n")
  cat(paste(" ", sort(profit_tags), collapse = "\n"), "\n")

  cat("\n--- 発行済株式数関連タグ候補 ---\n")
  cat(paste(" ", sort(shares_tags), collapse = "\n"), "\n")

  # --- 値抽出ヘルパー -------------------------------------------------------
  # context_prefer: contextRef に含まれるべき文字列（優先順）
  # context_exclude: contextRef に含まれていたら除外する文字列
  find_value <- function(keywords, context_prefer, context_exclude = character(0)) {

    node_to_record <- function(node) {
      txt <- trimws(xml_text(node))
      if (nchar(txt) == 0 || !grepl("^-?[0-9]+$", txt)) return(NULL)
      list(
        tag      = xml_name(node),
        value    = as.numeric(txt),
        context  = xml_attr(node, "contextRef"),
        decimals = xml_attr(node, "decimals"),
        unit_ref = xml_attr(node, "unitRef")
      )
    }

    for (kw in keywords) {
      xpath   <- sprintf("//*[contains(local-name(), '%s')]", kw)
      matches <- xml_find_all(doc, xpath)

      # 数値を持つ候補を全部収集
      candidates <- Filter(Negate(is.null), lapply(matches, node_to_record))
      if (length(candidates) == 0) next

      # 候補を全表示
      cat(sprintf("\n  [候補一覧: %s]\n", kw))
      for (r in candidates) {
        cat(sprintf("    contextRef=%-55s value=%s\n",
                    r$context, format(r$value, big.mark = ",")))
      }

      # 除外フィルタ
      if (length(context_exclude) > 0) {
        candidates <- Filter(
          function(r) !any(sapply(context_exclude, grepl, x = r$context)),
          candidates
        )
      }

      # 優先 contextRef を順番に試す
      for (pref in context_prefer) {
        hit <- Filter(function(r) grepl(pref, r$context), candidates)
        if (length(hit) > 0) return(hit[[1]])
      }

      # 優先パターンに合うものがなければ残った候補の先頭を返す
      if (length(candidates) > 0) return(candidates[[1]])
    }
    NULL
  }

  # --- 当期純利益 -----------------------------------------------------------
  cat("\n=== 抽出結果 ===\n")

  profit_priority <- c(
    "ProfitLossAttributableToOwnersOfParent",
    "NetIncomeLoss",
    "ProfitLoss",
    "NetIncome"
  )
  # CurrentYearDuration（連結・当期）を最優先。Prior系は除外しない（フォールバック用に残す）
  profit_context_prefer  <- c("CurrentYearDuration")
  profit_context_exclude <- character(0)
  res_profit <- find_value(profit_priority, profit_context_prefer, profit_context_exclude)

  if (!is.null(res_profit)) {
    cat(sprintf("\n当期純利益:\n"))
    cat(sprintf("  タグ名     : %s\n",   res_profit$tag))
    cat(sprintf("  contextRef : %s\n",   res_profit$context))
    cat(sprintf("  値（原単位）: %s\n",  format(res_profit$value, big.mark = ",")))
    cat(sprintf("  decimals   : %s\n",   res_profit$decimals))
    cat(sprintf("  unitRef    : %s\n",   res_profit$unit_ref))

    dec <- res_profit$decimals
    val <- res_profit$value
    if (!is.na(dec) && dec == "-6") {
      cat(sprintf("\n当期純利益：%s 百万円\n", format(val, big.mark = ",")))
    } else if (!is.na(dec) && dec == "0") {
      cat(sprintf("\n当期純利益：%s 百万円（%s 円換算）\n",
                  format(val / 1e6, big.mark = ","), format(val, big.mark = ",")))
    } else if (!is.na(dec) && dec == "-3") {
      cat(sprintf("\n当期純利益：%s 百万円（%s 千円換算）\n",
                  format(val / 1e3, big.mark = ","), format(val, big.mark = ",")))
    } else {
      cat(sprintf("\n当期純利益：%s（単位は decimals=%s を確認）\n",
                  format(val, big.mark = ","), dec))
    }
  } else {
    cat("\n当期純利益: タグが見つかりませんでした。\n")
    cat("上記「純利益関連タグ候補」からタグ名を確認してください。\n")
  }

  # --- 発行済株式数 ---------------------------------------------------------
  shares_priority <- c(
    "NumberOfIssuedSharesSummaryOfBusinessResults",
    "TotalNumberOfIssuedShares",
    "NumberOfIssuedShares",
    "IssuedShares"
  )
  # CurrentYearInstant を優先。"_NonConsolidatedMember" サフィックス付きは後回し
  shares_context_prefer  <- c("^CurrentYearInstant$", "CurrentYearInstant")
  shares_context_exclude <- character(0)
  res_shares <- find_value(shares_priority, shares_context_prefer, shares_context_exclude)

  if (!is.null(res_shares)) {
    cat(sprintf("\n発行済株式数:\n"))
    cat(sprintf("  タグ名     : %s\n",  res_shares$tag))
    cat(sprintf("  contextRef : %s\n",  res_shares$context))
    cat(sprintf("  値         : %s 株\n", format(res_shares$value, big.mark = ",")))
    cat(sprintf("\n発行済株式数：%s 株\n", format(res_shares$value, big.mark = ",")))
  } else {
    cat("\n発行済株式数: タグが見つかりませんでした。\n")
    cat("上記「発行済株式数関連タグ候補」からタグ名を確認してください。\n")
  }
}

# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------
check_api_key()

result      <- find_toyota_doc_id()
Sys.sleep(4)

zip_bytes   <- download_xbrl_zip(result$doc_id)
Sys.sleep(3)

xbrl_bytes  <- extract_xbrl_from_zip(zip_bytes)

parse_xbrl(xbrl_bytes)
